src/epic/distances/divergence_free.py

```diff
@@ -0,0 +1,217 @@
+                state_sampler is not None and action_sampler is not None
+            ), "If no coverage sampler is given, state and action samplers must be provided."
+        else:
+            assert (
+                state_sampler is None and action_sampler is None
+            ), "If a coverage sampler is given, state and action samplers will not be used."
+
+        coverage_sampler = coverage_sampler or samplers.ProductDistrCoverageSampler(action_sampler, state_sampler)
+
+        super().__init__(discount_factor, state_sampler, action_sampler, coverage_sampler)
+
+        state_sample, _, _, _ = self.coverage_sampler.sample(1)
+        self.state_dim = int(np.prod(state_sample.shape))
+
+    def init_potential_net(self):
+        """Initializes the neural net that'll be used to trained to strip potentials from reward functions during canonicalization."""
+
+    def canonicalize(
+        self,
+        reward_function: types.RewardFunction,
+        /,
+        n_samples_can: Optional[int],
+    ) -> types.RewardFunction:
+        """Canonicalizes a reward function into a divergence-free reward
+        function of the same equivalence class.
+
+        This is done by fitting a potential function (constructed from a neural
+        network) to minimize the L2 norm of the shaped reward function.
+        """
+        rew_fn = utils.multidim_rew_fn(reward_function)
+        n_samples_can = n_samples_can or self.default_samples_can
+        assert isinstance(n_samples_can, int)
+
+        net = nn.Sequential(
+            *[
+                nn.Flatten(),
+                nn.Linear(self.state_dim, max(self.state_dim * 4, 128)),
+                nn.GELU(),
+                Residual(nn.Linear(max(self.state_dim * 4, 128), max(self.state_dim * 4, 128))),
+                nn.GELU(),
+                nn.Linear(max(self.state_dim * 4, 128), 1),
+            ],
+        )
+
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        net.to(device)
+
+        max_epochs = 10000
+        optimizer = optim.AdamW(net.parameters(), lr=6e-4)
+        scheduler = optim.lr_scheduler.LambdaLR(
+            optimizer, lambda epoch: 0.5 if (epoch > 5000 and epoch < 7500) else 0.25 if (epoch > 7500) else 1.0
+        )
+
+        state_sample, action_sample, next_state_sample, done_sample = self.coverage_sampler.sample(
+            n_samples_can,
+        )
+
+        state_sample_tensor = utils.float_tensor_from_numpy(state_sample, device)
+        next_state_sample_tensor = utils.float_tensor_from_numpy(next_state_sample, device)
+
+        losses = []
+
+        with torch.inference_mode():
+            if state_sample_tensor.ndim == 1:
+                assert next_state_sample_tensor.ndim == 1
+                state_sample_tensor.unsqueeze_(-1)
+                next_state_sample_tensor.unsqueeze_(-1)
+
+            potential = (self.discount_factor * net(next_state_sample_tensor) - net(state_sample_tensor)).squeeze(-1)
```

Comment:
Note on terminology: "potential" means the function on states itself, i.e. the network in our case. This term with a difference of two potentials would be "potential shaping", "potential shaping term", "pure potential shaping", or maybe sometimes just "shaping term". I'd suggest thinking of it as a reward function that happens to have the special property of being induced by a potential. Mentioning that to avoid any confusion when I talk about "potentials" etc. (and for the same reason might be better to name this variable something like "shaping")

Reply:
Changed to `shaping`